### DISCUSSION TITLE
`belonging_to` query performance

### DISCUSSION THREAD

**Discussion by dractw** [diesel-rs/diesel#3188] [Q&A] [answered]
### Versions

- **Rust:** 1.59
- **Diesel:** 1.4.8
- **Database:** MySQL Percona cluster on InnoDB
- **Operating System** Debian

### Feature Flags

- **diesel:**

## Problem Description
Running `child::belonging_to(parent)`  on large dataset becomes slowly at scale

### Steps to reproduce

Schema looks like
```
table! {
    users (user_id) {
        id -> Unsigned<Integer>,
        user_id -> Varchar,
        app_id -> Varchar,
    }
}

table! {
    notes (id) {
        id -> Varchar,
        note_path -> Varchar,
        user_id -> Varchar,
    }
}

joinable!(notes -> users (user_id));

allow_tables_to_appear_in_same_query!(
    users,
    notes,
);
```

Each user have at least 5 notes, query at ~100.000 rows of users and ~500.000 rows of their notes and growing:
```
 let page_limit = u16::MAX / 2;
 let page_offset = *&page_limit as u32 * page_num as u32;
    
let app_users = users
    .filter(id.gt(page_offset).and(app_id.eq(note_app_id)))
    .limit(page_limit as i64)
    .load::<User>(&conn)?; // this query took ~0.26ms
    
let wallets: Vec<UserNotes> = UserNotes::belonging_to(&app_users)
              .filter(note_path.eq("default"))
              .load(&conn)?; // this query took ~8sec
```

Are there any workarounds to speedup notes query?

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by mohe2015** [comment]
I think this looks like a SQL optimization Problem. Could you maybe check the query that gets executed? And maybe get the query plan? I could imagine there are missing indexes or the database was not analyzed.

**Comment by weiznich** [comment, marked as answer]
I've converted your issue into a discussion forum thread, as I feel that it's more a question than an actual bug report. 

You have already noted that this particular query is slow.

```rust
let wallets: Vec<UserNotes> = UserNotes::belonging_to(&app_users)
              .filter(note_path.eq("default"))
              .load(&conn)?; // this query took ~8sec
```

This generates the following SQL query:
```sql
SELECT … FROM user_notes WHERE user_id IN (?, ?, …, ?) AND note_path = 'default';
```

where each entry in the bind list corresponds to one user id. That in turn means the issue here is likely a issue with your schema. As we don't have this information here, it's really hard to exactly tell you what needs to be done to improve this for you. As @mohe2015 already pointed out, looking into the generated query plan might help here. 

Things I personally would try to figure out why this is slow:

* Look at the query plan and try to understand where the time is spend.
* Try to add an index on the foreign key column
* Try to remove the extra filter and see if that has some effect.

**Reply by dractw** [reply]
Thank you for quick response!
Firstly i tried to remove filter ` .filter(note_path.eq("default"))` and response time became indecent ~80sec (10 times slower)

Additionaly, here's migrations schema:

```
CREATE TABLE IF NOT EXISTS `users` (
    `id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
    `user_id` varchar(255) NOT NULL,
    `app_id` varchar(255) NOT NULL,
    INDEX `id_in` (`id`),
    PRIMARY KEY (`user_id`)
);

CREATE TABLE IF NOT EXISTS `notes` (
    `id` varchar(255) NOT NULL,
    `note_path` varchar(255) NOT NULL,
    `user_id` varchar(255) NOT NULL,
    INDEX `note_path_in` (`note_path`),
    INDEX `notes_user_id_in` (`user_id`),
    PRIMARY KEY (`id`)
);

ALTER TABLE `notes` ADD CONSTRAINT `notes_fk0` FOREIGN KEY (`user_id`) REFERENCES `users`(`user_id`);
```

**Reply by mohe2015** [reply]
How is you data looking? How many app_users are you querying? How many notes does a user have? How many notes have a path of "default"? Are you sure the user_id is being indexed? Also you should really look at the query plan and the number of rows that are returned. I quickly tested this with sqlite3 and didn't see any obvious issue.
```sql
CREATE TABLE IF NOT EXISTS `users` (
    `id` INTEGER NOT NULL,
    `user_id` varchar(255) NOT NULL,
    `app_id` varchar(255) NOT NULL,
    PRIMARY KEY (`user_id`)
);
CREATE INDEX id_in ON users (id);

CREATE TABLE IF NOT EXISTS `notes` (
    `id` varchar(255) NOT NULL,
    `note_path` varchar(255) NOT NULL,
    `user_id` varchar(255) NOT NULL,
    PRIMARY KEY (`id`),
	FOREIGN KEY (`user_id`) REFERENCES `users`(`user_id`)
);
CREATE INDEX note_path_in ON notes (note_path);
CREATE INDEX notes_user_id_in ON notes (user_id);


INSERT INTO users
with recursive cte as (
      select random() as val, 1 as lev, random()
      union all
      select random(), lev + 1, random()
      from cte
      where lev < 100000
     )
select *
from cte;


-- execute 5 times
INSERT INTO notes
with recursive cte as (
      select random() as val, random(), 1 as lev
      union all
      select random(), random(), lev + 1
      from cte
      where lev < 100000
     )
select *
from cte;

EXPLAIN QUERY PLAN SELECT * FROM notes WHERE user_id IN (10, 215, 358, 325, 1335);

EXPLAIN QUERY PLAN SELECT * FROM notes WHERE user_id IN (10, 215, 358, 325, 1335) AND note_path = 'default';
```

**Reply by dractw** [reply]
As i remarked previously - issue came out only at large scaled dataset, so WHERE IN array is limited by  `u16::MAX / 2` at each paging

Using combined indexes as ```INDEX `user_notes` (`user_id`, `note_path`)``` did't help much. 

Here's `EXPLAIN ANALYZE SELECT * FROM notes WHERE user_id IN (SELECT user_id FROM users WHERE app_id = "some_app") AND note_path = "default"` output:
```
 -> Nested loop inner join  (cost=4616.75 rows=2855) (actual time=0.771..5585.267 rows=28067 loops=1)
    -> Filter: (users.app_id = 'some_app')  (cost=2943.36 rows=2774) (actual time=0.375..79.769 rows=28125 loops=1)
        -> Table scan on users  (cost=2943.36 rows=27743) (actual time=0.369..65.107 rows=28125 loops=1)
    -> Index lookup on notes using user_notes (user_id=users.user_id, note_path='default')  (cost=0.50 rows=1) (actual time=0.193..0.195 rows=1 loops=28125)
```

So there's ~28k users and ~565k notes, approx ~20 notes per user with different `note_path` (28k notes per 28k users with `default` note_path)

**Reply by mohe2015** [reply]
Guessing more or less blindly but I think you're missing an index on app_id. Also don't forget to reanalyze the database after index changes. Also it seems like there are 28125 users with that app_id so the necessarily need to be searched afterwards with that query design.
```sql
EXPLAIN QUERY PLAN SELECT * FROM notes WHERE user_id IN (SELECT user_id FROM users WHERE app_id = "some_app") AND note_path = "default"
ody>

3 | 0 | 0 | SEARCH notes USING INDEX note_path_in (note_path=?)
11 | 0 | 0 | LIST SUBQUERY 1
13 | 11 | 0 | SCAN users


CREATE INDEX notes_app_id ON users (app_id);


3 | 0 | 0 | SEARCH notes USING INDEX note_path_in (note_path=?)
11 | 0 | 0 | LIST SUBQUERY 1
14 | 11 | 0 | SEARCH users USING INDEX notes_app_id (app_id=?)
```

**Reply by mohe2015** [reply]
Ahh wait I looked at your code above again and you said that this wasn't the issue

**Reply by dractw** [reply]
@mohe2015 Thank you, will try this and share results.

**Reply by dractw** [reply]
Well, actually yes, selecting users by `app_id` runs fast. Seems like `WHERE IN` statement at large array is the key.

**Reply by mohe2015** [reply]
> Well, actually yes, selecting users by `app_id` runs fast. Seems like `WHERE IN` statement at large array is the key.

I also tried a few things out and I think it may be possible that with that design you're reaching the technical limits of the database. You may try a HASH index for user_id,note_path but other than that I think only putting the app_id into the notes table may help (which is a really bad idea because of normalization). Because then I think with a good index this could create a really localized query.

**Reply by mohe2015** [reply]
A few other things that come to my mind:
VARCHAR is probably less efficient than INTEGER or so for the user_id. So if this can be avoided this is probably better. Also the two id's for users is weird I think.

**Reply by mohe2015** [reply]
A few more things:

Maybe doing a JOIN instead of a IN query. I think pagination then is hard though. Also maybe the default notes can be put into a different database depending on what they're for as then all these other notes don't need to be searched (though an index should also work for that).

The pagination needs an order_by btw otherwise it's non-deterministic.

**Reply by mohe2015** [reply]
Wait now that I'm thinking more about this. Shouldn't the primary key of users be (app_id, user_id) and then notes contain these as foreign keys? Then I think this should get significantly faster with a JOIN

**Reply by mohe2015** [reply]
Or rather I think you don't need a join then any more.

Edit: For this combination of id and user_id you still need it but I think this design should be avoided.

**Reply by dractw** [reply]
Thank you for driving in this problem
Now im thinking that users table seems useless and just creates limitations. Will try to rewrite schema.

**Reply by mohe2015** [reply]
> Thank you for driving in this problem Now im thinking that users table seems useless and just creates limitations. Will try to rewrite schema.

I think currently it is indeed of not much use. But maybe if you later want to add a username or something you still want to keep it. But with a different schema layout it could be possible that it only rarely needs to be accessed which would probably be fine.
